File: src/sarma_cli/runtime/services.py
"""Runtime services passed into LangChain/LangGraph agent builders."""

from __future__ import annotations

from dataclasses import dataclass, field
from typing import Any


@dataclass
class AgentRuntimeServices:
    """Session-owned LangChain runtime services.

    Sarma's durable conversation history remains in ``sarma_cli.store.Store``.
    These services are scoped to the live runtime and are recreated on runtime
    restart.
    """

    checkpointer: Any
    store: Any
    cache: Any | None = None
    transformers: tuple[Any, ...] = field(default_factory=tuple)

    @classmethod
    def create(cls) -> "AgentRuntimeServices":
        from langgraph.checkpoint.memory import InMemorySaver
        from langgraph.store.memory import InMemoryStore

        return cls(
            checkpointer=InMemorySaver(),
            store=InMemoryStore(),
            # Cache is intentionally disabled until target-binary identity and
            # mutable IDA state are part of cache keys.
            cache=None,
            transformers=(),
        )

    def create_agent_kwargs(self) -> dict[str, Any]:
        kwargs: dict[str, Any] = {
            "checkpointer": self.checkpointer,
            "store": self.store,
        }
        if self.cache is not None:
            kwargs["cache"] = self.cache
        if self.transformers:
            kwargs["transformers"] = self.transformers
        return kwargs

    def compile_kwargs(self) -> dict[str, Any]:
        kwargs: dict[str, Any] = {
            "checkpointer": self.checkpointer,
            "store": self.store,
        }
        if self.cache is not None:
            kwargs["cache"] = self.cache
        if self.transformers:
            kwargs["transformers"] = self.transformers
        return kwargs

File: src/runtime/services.ts
/** Runtime services passed into LangChain/LangGraph agent builders. */

import { MemorySaver, InMemoryStore } from "@langchain/langgraph";
import type { BaseCheckpointSaver } from "@langchain/langgraph";

/**
 * Session-owned LangChain runtime services.
 *
 * Sarma's durable conversation history remains in the Store. These services
 * are scoped to the live runtime and are recreated on runtime restart.
 */
export class AgentRuntimeServices {
  readonly checkpointer: BaseCheckpointSaver;
  readonly store: InMemoryStore;
  readonly cache: unknown | null;

  constructor(
    checkpointer: BaseCheckpointSaver,
    store: InMemoryStore,
    cache: unknown | null = null,
  ) {
    this.checkpointer = checkpointer;
    this.store = store;
    this.cache = cache;
  }

  static create(): AgentRuntimeServices {
    return new AgentRuntimeServices(
      new MemorySaver(),
      new InMemoryStore(),
      // Cache is intentionally disabled until target-binary identity and
      // mutable IDA state are part of cache keys.
      null,
    );
  }

  /** Options passed to a compiled graph's invoke/stream config. */
  compileKwargs(): { checkpointer: BaseCheckpointSaver; store: InMemoryStore; cache?: unknown } {
    const kwargs: { checkpointer: BaseCheckpointSaver; store: InMemoryStore; cache?: unknown } = {
      checkpointer: this.checkpointer,
      store: this.store,
    };
    if (this.cache !== null) kwargs.cache = this.cache;
    return kwargs;
  }

  /** Options passed to `createAgent` (checkpointer + store + optional cache). */
  createAgentKwargs(): { checkpointer: BaseCheckpointSaver; store: InMemoryStore; cache?: unknown } {
    return this.compileKwargs();
  }
}
